### PR TITLE
[SPARK-11096] Post-hoc review Netty based RPC implementation - round 2

### PR DESCRIPTION
**Pull request by rxin**
A few more changes:
1. Renamed IDVerifier -> RpcEndpointVerifier
2. Renamed NettyRpcAddress -> RpcEndpointAddress
3. Simplified NettyRpcHandler a bit by removing the connection count tracking. This is OK because I now force spark.shuffle.io.numConnectionsPerPeer to 1
4. Reduced spark.rpc.connect.threads to 64. It would be great to eventually remove this extra thread pool.
5. Minor cleanup & documentation.

### CODE REVIEW

**Comment by rxin** [comment]
cc @zsxwing

File: core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala [hunk 1]
```diff
@@ -94,15 +94,6 @@ private[spark] abstract class RpcEnv(conf: SparkConf) {
   }
 
   /**
-   * Retrieve the [[RpcEndpointRef]] represented by `systemName`, `address` and `endpointName`
-   * asynchronously.
-   */
-  def asyncSetupEndpointRef(
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala]: this is not used anywhere

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala [hunk 2]
```diff
@@ -406,49 +409,37 @@ private[netty] class NettyRpcHandler(
   private type RemoteEnvAddress = RpcAddress
 
   // Store all client addresses and their NettyRpcEnv addresses.
+  // TODO: Is this even necessary?
   @GuardedBy("this")
   private val remoteAddresses = new mutable.HashMap[ClientAddress, RemoteEnvAddress]()
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala]: @zsxwing is this necessary at all? Maybe it is ok to not have this and always reconstruct the RpcAddress.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala]: Yes. It's necessary. `RemoteEnvAddress` is the port that a `RpcEnv` listens to, but `ClientAddress` is a random port that the client opens. For these network events, `RpcEndpoint` needs to use `RemoteEnvAddress`.

**Comment by zsxwing** [comment]
LGTM

**Comment by zsxwing** [comment]
Oh, `NettyRpcHandlerSuite.receive` tests the case that a remote client fires multiple connections. You can remove L45 and L46 here: https://github.com/apache/spark/blob/390b22fad69a33eb6daee25b6b858a2e768670a5/core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcHandlerSuite.scala#L45

**Comment by rxin** [comment]
Merging this .

**Pull request closed without merging**